fdb-relational-core/src/main/java/com/apple/foundationdb/relational/recordlayer/query/LiteralsUtils.java

```diff
@@ -55,25 +50,18 @@ private LiteralsUtils() {
 
     @Nonnull
     public static Type.Array resolveArrayTypeFromObjectsList(List<Object> objects) {
-        final var distinctTypes = objects.stream().filter(Objects::nonNull).map(SqlTypeSupport::getSqlTypeCodeFromObject).distinct().collect(Collectors.toList());
-        if (distinctTypes.isEmpty()) {
-            // has all nulls or is empty
-            return new Type.Array(Type.nullType());
-        }
-        Assert.thatUnchecked(distinctTypes.size() == 1, DATATYPE_MISMATCH, "could not determine type of array literal");
-        final var theType = distinctTypes.get(0);
-        if (theType != Types.STRUCT) {
-            return new Type.Array(Type.primitiveType(SqlTypeSupport.sqlTypeToRecordType(theType), false));
-        }
-        final var distinctStructMetadata = objects.stream().filter(Objects::nonNull).map(o -> {
-            try {
-                return ((RelationalStruct) o).getMetaData();
-            } catch (SQLException e) {
-                throw new RelationalException(e).toUncheckedWrappedException();
+        DataType distinctType = null;
+        for (var object: objects) {
```

Comment:
This is not entirely correct I think in the presence of nullable fields.

Reply:
Opened issue on the same: https://github.com/FoundationDB/fdb-record-layer/issues/3472